### ISSUE TITLE
Release Securedrop 2.9.0

### ISSUE THREAD

**Issue by zenmonkeykstop** [freedomofpress/securedrop#7168]
This is a tracking issue for the release of SecureDrop 2.9.0

Tentatively scheduled as follows:

**Pre-release announcement:** TK
**Release date:** TK (no later than 2024-07-05)

**Release manager:** @zenmonkeykstop  
**Deputy release manager:** @legoktm 
**Localization manager:** n/a (no string changes)
**Communications manager:** @nathandyer 

_SecureDrop maintainers and testers:_ As you QA 2.9.0, please report back your testing results as comments on this ticket. File GitHub issues for any problems found, tag them "QA: Release".

Test debian packages will be posted on https://apt-test.freedom.press signed with [the test key](https://gist.githubusercontent.com/conorsch/ec4008b111bc3142fca522693f3cce7e/raw/2968621e8ad92db4505a31fcc5776422d7d26729/apt-test%2520apt%2520pubkey).

# [QA Matrix for 2.9.0](https://docs.google.com/spreadsheets/d/1suShRhMPmYc9t7vGxifJ7AFP5ERwptP29JfPoXBEfMo/edit#gid=96348658)
# [Test Plan for 2.9.0](https://github.com/freedomofpress/securedrop/wiki/2.9.0-Test-Plan)

# Prepare release candidate (2.9.0~rc1)
- [x] Link to latest version of Tails, including release candidates, to test against during QA
- [x] Prepare 2.9.0~rc1 release changelog
- [x] Branch off release/2.9.0 from develop
- [x] Prepare 2.9.0-rc1
- [x] Build debs, preserving build log, and put up `2.9.0~rc1` on test apt server
- [x] Commit build log.

# Prepare release candidate (2.9.0~rc2)

- [x] Prepare 2.9.0~rc2 release changelog
- [x] Prepare 2.9.0-rc2
- [x] Build debs, preserving build log
- [x] Commit build log
- [x] Put up `2.9.0~rc2` on test apt server.

# Prepare release candidate (2.9.0~rc3)

- [x] Prepare 2.9.0~rc3 release changelog
- [x] Prepare 2.9.0-rc3
- [x] Build debs, preserving build log
- [x] Commit build log
- [ ] Put up `2.9.0~rc3` on test apt server.

After each test, please update the QA matrix and post details for Basic Server Testing, Application Acceptance Testing and release-specific testing below in comments to this ticket.

# Final release
- [ ] ~Ensure builder in release branch is updated and/or update builder image~ (no longer in use)
- [x] Push signed tag 
- [x] Pre-Flight: Test updater logic in Tails (apt-qa tracks the `release` branch in the LFS repo)
- [x] Build final Debian packages(and preserve build log)
- [x] Commit package build log to https://github.com/freedomofpress/build-logs
- [x] Pre-Flight: Test that install and upgrade from 2.8.0 to 2.9.0 works w/ prod repo debs (apt-qa.freedom.press polls the `release` branch in the LFS repo for the debs)
- [x] Flip apt QA server to prod status (merge to `main` in the LFS repo)
- [x] Merge Docs branch changes to ``main`` and verify new docs build in securedrop-docs repo
- [x] Prepare release messaging

# Post release
- [x] Create GitHub release object 
- [x] Once release object is created, update versions in `securedrop-docs` and Wagtail
- [x] Verify new docs show up on https://docs.securedrop.org
- [x] Publish announcements
- [ ] Merge changelog back to `develop`
- [x] Update roadmap wiki page: https://github.com/freedomofpress/securedrop/wiki/Development-Roadmap

**Comment by nathandyer** [comment]
### Environment

- Install target: NUC13 (App) and NUC12 (Mon)
- Tails version: 6.3
- Test Scenario: Upgrade
- SSH over Tor: No
- Release candidate: 2.9.0~rc1
- General notes:

### Basic Server Testing
- [x] After installing the testinfra dependencies, all tests in `./securedrop-admin verify` are passing:
  - Install dependencies on Admin Workstation with `cd ~/Persistent/securedrop && ./securedrop-admin setup -t`
  - Run tests with `./securedrop-admin verify` (this will take a while)
  - Remove test dependencies: `rm -rf admin/.venv3/ && ./securedrop-admin setup`
- [x] QA Matrix checks pass

#### Command Line User Generation

- [x] Can successfully add admin user and login

#### (Optional) Administration

- [ ] I have backed up and successfully restored the app server following [the backup documentation](https://docs.securedrop.org/en/latest/backup_and_restore.html)
- [ ] If doing upgrade testing, make a backup on 2.8.0 and restore this backup on this release candidate
- [x] "Send Test OSSEC Alert" button in the journalist triggers an OSSEC alert and an email is sent
- [ ] Can successfully add journalist account with HOTP authentication

### (Optional) Application Acceptance Testing

#### Source Interface

##### Landing page base cases
- [x] JS warning bar does not appear when using Security Slider high
- [x] JS warning bar does appear when using Security Slider Low

##### First submission base cases
- [x] On generate page, refreshing page produces a new 7-word codename
- [x] On submit page, empty submissions produce flashed message
- [x] On submit page, short message submitted successfully
- [x] On submit page, file greater than 500 MB produces "The connection was reset" in Tor Browser _quickly_ before the entire file is uploaded
- [x] On submit page, file less than 500 MB submitted successfully

##### Returning source base cases

- [x] Nonexistent codename cannot log in
- [x] Empty codename cannot log in
- [x] Legitimate codename can log in
- [x] Returning user can view journalist replies - need to log into journalist interface to test

#### Journalist Interface

##### Login base cases
- [x] Can log in with 2FA tokens
- [x] incorrect password cannot log in
- [x] invalid 2fa token cannot log in
- [x] 2fa immediate reuse cannot log in
- [ ] Journalist account with HOTP can log in

##### Index base cases

- [x] Filter by codename works
- [x] Starring and unstarring works
- [ ] Click select all selects all submissions
- [ ] Selecting all and clicking "Download" works

##### Individual source page

- [x] You can submit a reply and a flashed message and new row appears
- [x] You cannot submit an empty reply
- [x] Clicking "Delete Source Account" and the source and docs are deleted
- [x] You can click on a document and successfully decrypt using application private key

### Basic Tails Testing

After updating to this release candidate and running `securedrop-admin tailsconfig`

- [x] The Updater GUI appears on boot

### 2.9.0 release-specific changes

#### Update release key expiry [#7167](https://github.com/freedomofpress/securedrop/pull/7167)
- [x] after installation or upgrade, verify that the key in /etc/apt/trusted.gpg.d/securedrop-keyring.gpg is the SecureDrop signing key with an expiry of 207-05024

#### API range request support [#7160](https://github.com/freedomofpress/securedrop/pull/7160)
- set up a journalist user via the JI
- install jq in Tails with `sudo apt install jq`
- From a terminal, get an API token with a command like:
  `curl -X POST -H "Content-Type: application/json" -d '{"username": "<journalist_name>", "passphrase": "<passphrase>", "one_time_code": "<code"}' http://<JI address>/api/v1/token`
- extract the token from the response and store it in an env var, e.g `TOKEN=<token_string>`
- Get the submission and download URL for a file submission:
  ```
  SUBMISSION_URL=$(curl -H "Authorization: Token $TOKEN" http://<JI address>/api/v1/sources | jq -r '.sources[-1].submissions_url')
  DOWNLOAD_URL=$(curl -H "Authorization: Token $TOKEN" http://<JI address>$SUBMISSION_URL | jq -r '.submissions[0].download_url')
  ```
- [x] verify that you can download the full submission: `curl -v -H "Authorization: Token $TOKEN"  -o logo.png.gpg http://<JI_address>$DOWNLOAD_URL`
- [x] verify that you can download the first 100 bytes of the submission: `curl -v -H "Authorization: Token $TOKEN" -H "Range: bytes=0-99" -o logo.png.gpg.part-0-99 http://l<JI_address>$DOWNLOAD_URL`
- [x] Verify that the first 100 bytes of the original submission has the same sha256 hash as the partial download.


### Preflight testing

#### Basic testing

- [x] Install or upgrade occurs without error (from `apt-qa.freedom.press` per [preflight procedure](https://github.com/freedomofpress/securedrop/wiki/QA-Procedures#pre-release-prep))
- [x] Source interface is available and version string indicates it is 2.9.0
- [x] A message can be successfully submitted

#### Tails

- [x] The updater GUI appears on boot
- [x] The update successfully occurs to 2.9.0
- [x] After reboot, updater GUI no longer appears

**Comment by nathandyer** [comment]
**In Progress**
### Environment

- Install target: NUC13 (App) and NUC12 (Mon)
- Tails version: 6.3
- Test Scenario: Install
- SSH over Tor: No
- Release candidate: 2.9.0~rc1
- General notes:

### Basic Server Testing
- [x] After installing the testinfra dependencies, all tests in `./securedrop-admin verify` are passing:
  - Install dependencies on Admin Workstation with `cd ~/Persistent/securedrop && ./securedrop-admin setup -t`
  - Run tests with `./securedrop-admin verify` (this will take a while)
  - Remove test dependencies: `rm -rf admin/.venv3/ && ./securedrop-admin setup`
- [x] QA Matrix checks pass

#### Command Line User Generation

- [x] Can successfully add admin user and login

#### (Optional) Administration

- [ ] I have backed up and successfully restored the app server following [the backup documentation](https://docs.securedrop.org/en/latest/backup_and_restore.html)
- [ ] If doing upgrade testing, make a backup on 2.8.0 and restore this backup on this release candidate
- [x] "Send Test OSSEC Alert" button in the journalist triggers an OSSEC alert and an email is sent
- [ ] Can successfully add journalist account with HOTP authentication

### (Optional) Application Acceptance Testing

#### Source Interface

##### Landing page base cases
- [x] JS warning bar does not appear when using Security Slider high
- [x] JS warning bar does appear when using Security Slider Low

##### First submission base cases
- [x] On generate page, refreshing page produces a new 7-word codename
- [x] On submit page, empty submissions produce flashed message
- [x] On submit page, short message submitted successfully
- [x] On submit page, file greater than 500 MB produces "The connection was reset" in Tor Browser _quickly_ before the entire file is uploaded
- [x] On submit page, file less than 500 MB submitted successfully

##### Returning source base cases

- [x] Nonexistent codename cannot log in
- [x] Empty codename cannot log in
- [x] Legitimate codename can log in
- [x] Returning user can view journalist replies - need to log into journalist interface to test

#### Journalist Interface

##### Login base cases
- [x] Can log in with 2FA tokens
- [x] incorrect password cannot log in
- [x] invalid 2fa token cannot log in
- [x] 2fa immediate reuse cannot log in
- [ ] Journalist account with HOTP can log in

##### Index base cases

- [x] Filter by codename works
- [x] Starring and unstarring works
- [ ] Click select all selects all submissions
- [ ] Selecting all and clicking "Download" works

##### Individual source page

- [x] You can submit a reply and a flashed message and new row appears
- [x] You cannot submit an empty reply
- [x] Clicking "Delete Source Account" and the source and docs are deleted
- [x] You can click on a document and successfully decrypt using application private key

### Basic Tails Testing

After updating to this release candidate and running `securedrop-admin tailsconfig`

- [x] The Updater GUI appears on boot

### 2.9.0 release-specific changes

#### Update release key expiry [#7167](https://github.com/freedomofpress/securedrop/pull/7167)
- [x] after installation or upgrade, verify that the key in /etc/apt/trusted.gpg.d/securedrop-keyring.gpg is the SecureDrop signing key with an expiry of 207-05024

#### API range request support [#7160](https://github.com/freedomofpress/securedrop/pull/7160)
- set up a journalist user via the JI
- install jq in Tails with `sudo apt install jq`
- From a terminal, get an API token with a command like:
  `curl -X POST -H "Content-Type: application/json" -d '{"username": "<journalist_name>", "passphrase": "<passphrase>", "one_time_code": "<code"}' http://<JI address>/api/v1/token`
- extract the token from the response and store it in an env var, e.g `TOKEN=<token_string>`
- Get the submission and download URL for a file submission:
  ```
  SUBMISSION_URL=$(curl -H "Authorization: Token $TOKEN" http://<JI address>/api/v1/sources | jq -r '.sources[-1].submissions_url')
  DOWNLOAD_URL=$(curl -H "Authorization: Token $TOKEN" http://<JI address>$SUBMISSION_URL | jq -r '.submissions[0].download_url')
  ```
- [x] verify that you can download the full submission: `curl -v -H "Authorization: Token $TOKEN"  -o logo.png.gpg http://<JI_address>$DOWNLOAD_URL`
- [ ] verify that you can download the first 100 bytes of the submission: `curl -v -H "Authorization: Token $TOKEN" -H "Range: bytes=0-99" -o logo.png.gpg.part-0-99 http://l<JI_address>$DOWNLOAD_URL`
- [ ] Verify that the first 100 bytes of the original submission has the same sha256 hash as the partial download.


### Preflight testing

#### Basic testing

- [x] Install or upgrade occurs without error (from `apt-qa.freedom.press` per [preflight procedure](https://github.com/freedomofpress/securedrop/wiki/QA-Procedures#pre-release-prep))
- [x] Source interface is available and version string indicates it is 2.9.0
- [x] A message can be successfully submitted

#### Tails

- [x] The updater GUI appears on boot
- [x] The update successfully occurs to 2.9.0
- [x] After reboot, updater GUI no longer appears

**Comment by cfm** [comment]
So far I've tested *only* the Tor proof-of-work defense (on and back off) on the upgrade path.  @zenmonkeykstop, we can check in tomorrow about how much more general testing you'd like on this scenario.


### Environment

- Install target: VMs
- Tails version: 6.4
- Test Scenario: upgrade
- SSH over Tor?: yes
- Tor PoW enabled?: off, on, off
- Release candidate: `2.9.0-rc2`
- General notes:

### Basic Server Testing
- [x] After installing the testinfra dependencies, all tests in `./securedrop-admin verify` are passing:
  - Install dependencies on Admin Workstation with `cd ~/Persistent/securedrop && ./securedrop-admin setup -t`
  - Run tests with `./securedrop-admin verify` (this will take a while)
  - Remove test dependencies: `rm -rf admin/.venv3/ && ./securedrop-admin setup`
- [ ] QA Matrix checks pass

#### Command Line User Generation

- [ ] Can successfully add admin user and login

#### (Optional) Administration

- [x] I have backed up and successfully restored the app server following [the backup documentation](https://docs.securedrop.org/en/latest/backup_and_restore.html)
- [x] If doing upgrade testing, make a backup on 2.8.0 and restore this backup on this release candidate
- [ ] "Send Test OSSEC Alert" button in the journalist triggers an OSSEC alert and an email is sent
- [ ] Can successfully add journalist account with HOTP authentication

### (Optional) Application Acceptance Testing

#### Source Interface

##### Landing page base cases
- [ ] JS warning bar does not appear when using Security Slider high
- [ ] JS warning bar does appear when using Security Slider Low

##### First submission base cases
- [ ] On generate page, refreshing page produces a new 7-word codename
- [ ] On submit page, empty submissions produce flashed message
- [ ] On submit page, short message submitted successfully
- [ ] On submit page, file greater than 500 MB produces "The connection was reset" in Tor Browser _quickly_ before the entire file is uploaded
- [ ] On submit page, file less than 500 MB submitted successfully

##### Returning source base cases

- [ ] Nonexistent codename cannot log in
- [ ] Empty codename cannot log in
- [ ] Legitimate codename can log in
- [ ] Returning user can view journalist replies - need to log into journalist interface to test

#### Journalist Interface

##### Login base cases
- [ ] Can log in with 2FA tokens
- [ ] incorrect password cannot log in
- [ ] invalid 2fa token cannot log in
- [ ] 2fa immediate reuse cannot log in
- [ ] Journalist account with HOTP can log in

##### Index base cases

- [ ] Filter by codename works
- [ ] Starring and unstarring works
- [ ] Click select all selects all submissions
- [ ] Selecting all and clicking "Download" works

##### Individual source page

- [ ] You can submit a reply and a flashed message and new row appears
- [ ] You cannot submit an empty reply
- [ ] Clicking "Delete Source Account" and the source and docs are deleted
- [ ] You can click on a document and successfully decrypt using application private key

### Basic Tails Testing

After updating to this release candidate and running `securedrop-admin tailsconfig`

- [ ] The Updater GUI appears on boot

### 2.9.0 release-specific changes

#### Tor PoW defense [#7175](https://github.com/freedomofpress/securedrop/pull/7175)
- Install `onionprobe` on the admin workstation if not already present

##### Upgrade
- after completing the upgrade procedure on the servers and admin workstation:
- [x] run `./securedrop-admin sdconfig` and confirm that the option to enable PoW defense is available - choose to enable it
- run `./securedrop-admin install` and confirm:
  - [x] all Tor hidden services are available
  - [x] `onionprobe -e <si address>` confirms that PoW is enabled for the source interface

- [x] run `./securedrop-admin sdconfig` again and disable PoW defense
- run `./securedrop-admin install` and confirm:
  - [x] all Tor hidden services are available
  - [ ] `onionprobe -e <SI address>` confirms that PoW is *not* enabled for the source interface

#### Update release key expiry [#7167](https://github.com/freedomofpress/securedrop/pull/7167)
- [x] after installation or upgrade, verify that the key in /etc/apt/trusted.gpg.d/securedrop-keyring.gpg is the SecureDrop signing key with an expiry of 2027-05-24

#### API range request support [#7160](https://github.com/freedomofpress/securedrop/pull/7160)
- set up a journalist user via the JI
- install jq in Tails with `sudo apt install jq`
- From a terminal, get an API token with a command like:
  `curl -X POST -H "Content-Type: application/json" -d '{"username": "<journalist_name>", "passphrase": "<passphrase>", "one_time_code": "<code"}' http://<JI address>/api/v1/token`
- extract the token from the response and store it in an env var, e.g `TOKEN=<token_string>`
- Get the submission and download URL for a file submission:
  ```
  SUBMISSION_URL=$(curl -H "Authorization: Token $TOKEN" http://<JI address>/api/v1/sources | jq -r '.sources[-1].submissions_url')
  DOWNLOAD_URL=$(curl -H "Authorization: Token $TOKEN" http://<JI address>$SUBMISSION_URL | jq -r '.submissions[0].download_url')
  ```
- [ ] verify that you can download the full submission: `curl -v -H "Authorization: Token $TOKEN"  -o logo.png.gpg http://<JI_address>$DOWNLOAD_URL`
- [ ] verify that you can download the first 100 bytes of the submission: `curl -v -H "Authorization: Token $TOKEN" -H "Range: bytes=0-99" -o logo.png.gpg.part-0-99 http://l/<JI_address>$DOWNLOAD_URL`
- [ ] Verify that the first 100 bytes of the original submission has the same sha256 hash as the partial download.

**Comment by zenmonkeykstop** [comment]
## 2.9.0 QA Checklist **IN PROGRESS


Note that this release includes changes to the server configuration to enable Tor PoW - you should complete the relevant test plan section while performing the installation or upgrade.

### Environment

- Install target: VM
- Tails version: 6.3
- Test Scenario: fresh install
- SSH over Tor?: yup
- Tor PoW enabled?: yup
- Release candidate: 2.9.0-rc2
- General notes:

### Basic Server Testing
- [x] After installing the testinfra dependencies, all tests in `./securedrop-admin verify` are passing:
  - Install dependencies on Admin Workstation with `cd ~/Persistent/securedrop && ./securedrop-admin setup -t`
  - Run tests with `./securedrop-admin verify` (this will take a while)
  - Remove test dependencies: `rm -rf admin/.venv3/ && ./securedrop-admin setup`
- [x] QA Matrix checks pass

#### Command Line User Generation

- [x] Can successfully add admin user and login

#### (Optional) Administration

- [ ] I have backed up and successfully restored the app server following [the backup documentation](https://docs.securedrop.org/en/latest/backup_and_restore.html)
- [ ] ~If doing upgrade testing, make a backup on 2.8.0 and restore this backup on this release candidate~ skipped
- [x] "Send Test OSSEC Alert" button in the journalist triggers an OSSEC alert and an email is sent
- [ ] ~Can successfully add journalist account with HOTP authentication~ skipped

### (Optional) Application Acceptance Testing **SKIPPED**


### Basic Tails Testing

After updating to this release candidate and running `securedrop-admin tailsconfig`

- [x] The Updater GUI appears on boot

### 2.9.0 release-specific changes

#### Tor PoW defense [#7175](https://github.com/freedomofpress/securedrop/pull/7175)

##### Fresh install
- [x] when running `./securedrop-admin sdconfig`, confirm that the option to enable PoW defense is available - choose to enable it
- run `./securedrop-admin install` and `./securedrop-admin tailsconfig and confirm:
  - [x] both commands completed successfully
  - [x] all Tor hidden services are available
  - [x] `onionprobe -e <si address>` (on a remote machine) confirms that PoW is enabled for the source interface 

##### Upgrade *N/A*

#### Update release key expiry [#7167](https://github.com/freedomofpress/securedrop/pull/7167)
- [x] after installation or upgrade, verify that the key in /etc/apt/trusted.gpg.d/securedrop-keyring.gpg is the SecureDrop signing key with an expiry of 2027-05-24

#### API range request support [#7160](https://github.com/freedomofpress/securedrop/pull/7160)
- set up a journalist user via the JI
- install jq in Tails with `sudo apt install jq`
- From a terminal, get an API token with a command like:
  `curl -X POST -H "Content-Type: application/json" -d '{"username": "<journalist_name>", "passphrase": "<passphrase>", "one_time_code": "<code"}' http://<JI address>/api/v1/token`
- extract the token from the response and store it in an env var, e.g `TOKEN=<token_string>`
- Get the submission and download URL for a file submission:
  ```
  SUBMISSION_URL=$(curl -H "Authorization: Token $TOKEN" http://<JI address>/api/v1/sources | jq -r '.sources[-1].submissions_url')
  DOWNLOAD_URL=$(curl -H "Authorization: Token $TOKEN" http://<JI address>$SUBMISSION_URL | jq -r '.submissions[0].download_url')
  ```
- [x] verify that you can download the full submission: `curl -v -H "Authorization: Token $TOKEN"  -o logo.png.gpg http://<JI_address>$DOWNLOAD_URL`
- [ ] verify that you can download the first 100 bytes of the submission: `curl -v -H "Authorization: Token $TOKEN" -H "Range: bytes=0-99" -o logo.png.gpg.part-0-99 http://l<JI_address>$DOWNLOAD_URL` **FAIL** see https://github.com/freedomofpress/securedrop/issues/7186
- [ ] Verify that the first 100 bytes of the original submission has the same sha256 hash as the partial download. **FAIL**


### Preflight testing

#### Basic testing

- [ ] Install or upgrade occurs without error (from `apt-qa.freedom.press` per [preflight procedure](https://github.com/freedomofpress/securedrop/wiki/QA-Procedures#pre-release-prep))
- [ ] Source interface is available and version string indicates it is 2.9.0
- [ ] A message can be successfully submitted

#### Tails

- [ ] The updater GUI appears on boot
- [ ] The update successfully occurs to 2.9.0
- [ ] After reboot, updater GUI no longer appears

**Comment by nathandyer** [comment]
Just repeated the range request tests on rc2 after reports that it wasn't working for some folks. I can now confirm that the range requests are _not_ working for me, either on rc2 or when I manually fall back to rc1. 

I receive: `Closing connection 0; transfer closed with 100 bytes remaining to read`

Not sure why it worked for me a couple weeks ago (I'm questioning if I was just mistaken, although I remember having to look up how to carve out the first 100 bytes of `logo.png.gpg` so I could compare it against `logo.png.gpg.part0-99`, which makes me think I really was able to do it on my original attempt).

**Comment by nathandyer** [comment]
### Environment

- Install target: NUC13 (App) and NUC12 (Mon)
- Tails version: 6.3
- Test Scenario: Install
- SSH over Tor?: No
- Tor PoW enabled?: Yes
- Release candidate: rc3
- General notes:


### 2.9.0 release-specific changes

#### Tor PoW defense [#7175](https://github.com/freedomofpress/securedrop/pull/7175)
- Install `onionprobe` on the admin workstation if not already present

##### Fresh install
- [x] when running `./securedrop-admin sdconfig`, confirm that the option to enable PoW defense is available - choose to enable it
- run `./securedrop-admin install` and `./securedrop-admin tailsconfig and confirm:
  - [x] both commands completed successfully
  - [x] all Tor hidden services are available
  - [x] `onionprobe -e <si address>` confirms that PoW is enabled for the source interface 
<details>

```
2024-06-26 15:24:15,737 INFO: Proof of Work (PoW) params found in the descriptor
2024-06-26 15:24:15,739 INFO: PoW v1 set with effort 0, expiration 2024-06-26T21:12:27 and seed [redacted]
```
</details>

#### Update release key expiry [#7167](https://github.com/freedomofpress/securedrop/pull/7167)
- [x] after installation or upgrade, verify that the key in /etc/apt/trusted.gpg.d/securedrop-keyring.gpg is the SecureDrop signing key with an expiry of 2027-05-24

#### API range request support [#7160](https://github.com/freedomofpress/securedrop/pull/7160)
- set up a journalist user via the JI
- install jq in Tails with `sudo apt install jq`
- From a terminal, get an API token with a command like:
  `curl -X POST -H "Content-Type: application/json" -d '{"username": "<journalist_name>", "passphrase": "<passphrase>", "one_time_code": "<code"}' http://<JI address>/api/v1/token`
- extract the token from the response and store it in an env var, e.g `TOKEN=<token_string>`
- Get the submission and download URL for a file submission:
  ```
  SUBMISSION_URL=$(curl -H "Authorization: Token $TOKEN" http://<JI address>/api/v1/sources | jq -r '.sources[-1].submissions_url')
  DOWNLOAD_URL=$(curl -H "Authorization: Token $TOKEN" http://<JI address>$SUBMISSION_URL | jq -r '.submissions[0].download_url')
  ```
- [x] verify that you can download the full submission: `curl -v -H "Authorization: Token $TOKEN"  -o logo.png.gpg http://<JI_address>$DOWNLOAD_URL`
- [x] verify that you can download the first 100 bytes of the submission: `curl -v -H "Authorization: Token $TOKEN" -H "Range: bytes=0-99" -o logo.png.gpg.part-0-99 http://l<JI_address>$DOWNLOAD_URL`
- [x] Verify that the first 100 bytes of the original submission has the same sha256 hash as the partial download.

![Screenshot from 2024-06-26 19-43-53](https://github.com/freedomofpress/securedrop/assets/6131358/e9bc3f1d-9b00-47c5-9a5b-2b4ad5f7bf9d)

**Comment by zenmonkeykstop** [comment]
## 2.9.0-rc3 QA Checklist - **IN PROGRESS** 

For both upgrades and fresh installs, here is a list of functionality that requires testing. You can use this for copy/pasting into your QA report.

If you have submitted a QA report already for a release candidate with successful basic server testing and application acceptance testing sections, then you can skip these sections in subsequent reports, unless otherwise indicated by the Release Manager. This is to ensure that you focus your QA effort on the release-specific changes as well as changes since the previous release candidate.

Note that this release includes changes to the server configuration to enable Tor PoW - you should complete the relevant test plan section while performing the installation or upgrade.

### Environment

- Install target: VMs
- Tails version: 6.4
- Test Scenario: upgrade
- SSH over Tor?: No
- Tor PoW enabled?: Yes
- Release candidate: RC3
- General notes:

### Basic Server Testing
- [x] After installing the testinfra dependencies, all tests in `./securedrop-admin verify` are passing:
  - Install dependencies on Admin Workstation with `cd ~/Persistent/securedrop && ./securedrop-admin setup -t`
  - Run tests with `./securedrop-admin verify` (this will take a while)
  - Remove test dependencies: `rm -rf admin/.venv3/ && ./securedrop-admin setup`
- [x] QA Matrix checks pass

#### Command Line User Generation

- [ ] Can successfully add admin user and login

#### (Optional) Administration **SKIPPED**


### (Optional) Application Acceptance Testing **SKIPPED**

### 2.9.0 release-specific changes

#### Tor PoW defense [#7175](https://github.com/freedomofpress/securedrop/pull/7175)
- Install `onionprobe` on the admin workstation if not already present

##### Fresh install **N/A**

##### Upgrade
- after completing the upgrade procedure on the servers and admin workstation:
- [x] run `./securedrop-admin sdconfig` and confirm that the option to enable PoW defense is available - choose to enable it
- run `./securedrop-admin install` and confirm:
  - [x] all Tor hidden services are available
  - [x] `onionprobe -e <si address>` confirms that PoW is enabled for the source interface

- [x] run `./securedrop-admin sdconfig` again and disable PoW defense
- run `./securedrop-admin install` and confirm:
  - [x] all Tor hidden services are available
  - [x] `onionprobe -e <SI address>` confirms that PoW is *not* enabled for the source interface

#### Update release key expiry [#7167](https://github.com/freedomofpress/securedrop/pull/7167)
- [x] after installation or upgrade, verify that the key in /etc/apt/trusted.gpg.d/securedrop-keyring.gpg is the SecureDrop signing key with an expiry of 2027-05-24

#### API range request support [#7160](https://github.com/freedomofpress/securedrop/pull/7160)
- set up a journalist user via the JI
- install jq in Tails with `sudo apt install jq`
- From a terminal, get an API token with a command like:
  `curl -X POST -H "Content-Type: application/json" -d '{"username": "<journalist_name>", "passphrase": "<passphrase>", "one_time_code": "<code"}' http://<JI address>/api/v1/token`
- extract the token from the response and store it in an env var, e.g `TOKEN=<token_string>`
- Get the submission and download URL for a file submission:
  ```
  SUBMISSION_URL=$(curl -H "Authorization: Token $TOKEN" http://<JI address>/api/v1/sources | jq -r '.sources[-1].submissions_url')
  DOWNLOAD_URL=$(curl -H "Authorization: Token $TOKEN" http://<JI address>$SUBMISSION_URL | jq -r '.submissions[0].download_url')
  ```
- [x] verify that you can download the full submission: `curl -v -H "Authorization: Token $TOKEN"  -o logo.png.gpg http://<JI_address>$DOWNLOAD_URL`
- [x] verify that you can download the first 100 bytes of the submission: `curl -v -H "Authorization: Token $TOKEN" -H "Range: bytes=0-99" -o logo.png.gpg.part-0-99 http://l<JI_address>$DOWNLOAD_URL`
- [x] Verify that the first 100 bytes of the original submission has the same sha256 hash as the partial download.


### Preflight testing

#### Basic testing

- [ ] Install or upgrade occurs without error (from `apt-qa.freedom.press` per [preflight procedure](https://github.com/freedomofpress/securedrop/wiki/QA-Procedures#pre-release-prep))
- [ ] Source interface is available and version string indicates it is 2.9.0
- [ ] A message can be successfully submitted

#### Tails

- [ ] The updater GUI appears on boot
- [ ] The update successfully occurs to 2.9.0
- [ ] After reboot, updater GUI no longer appears

**Comment by zenmonkeykstop** [comment]
## install-scenario preflight:
- [x] Install or upgrade occurs without error (from apt-qa.freedom.press per [preflight procedure](https://github.com/freedomofpress/securedrop/wiki/QA-Procedures#pre-release-prep))
- [x] Source interface is available and version string indicates it is 2.9.0
- [x] A message can be successfully submitted

Tails

- [x] The updater GUI appears on boot
- [x] The update successfully occurs to 2.9.0
- [x] After reboot, updater GUI no longer appears

**Comment by cfm** [comment]
- Install target: VMs
- Tails version: 6.4
- Test Scenario: upgrade
- SSH over Tor?: yes
- Tor PoW enabled?: no
- Release candidate: `2.9.0`

### Preflight testing

#### Basic testing

- [x] Install or upgrade occurs without error (from `apt-qa.freedom.press` per [preflight procedure](https://github.com/freedomofpress/securedrop/wiki/QA-Procedures#pre-release-prep))
- [x] Source interface is available and version string indicates it is 2.9.0
- [x] A message can be successfully submitted

#### Tails

- [x] The updater GUI appears on boot
- [x] The update successfully occurs to 2.9.0

Some fancy footwork required for updating a shallow clone.  Documentation forthcoming.

- [x] After reboot, updater GUI no longer appears
